Comment router

a-w-m opened the pull request:
--Update: I updated the code given the comments and tested each route. I changed the status of this PR as "ready to review". It's no longer a draft.

----
Hey everyone. This is a draft PR (not meant to be merged) of my work so far on comment router. I made assumptions about the request.body in the functions, and haven't validated or sanitized. Since I'm new to the back-end I just wanted to share what I've come up with and get your feedback, guidance, and ensure I'm on/near the right track. Thanks!

a-w-m committed:
Creates comment router and router folder

a-w-m committed:
Merge branch 'dev' into comment-router

rusye commented on src/routers/comment-router.js:
```diff
@@ -0,0 +1,95 @@
+returns a promise of a JSON object containing a list of comments
+*/
+router.get("/", function(req, res) {
+  const { userID, username, text, date } = req.body;
```
We won't be passing ` userID, username, text, date ` in the req because we want to return everything.

rusye replied:
You should be able to just pass `Comment.find()`

a-w-m replied:
> You should be able to just pass `Comment.find()`

Okay, I think I got it. So, Comment.find() with no search argument. And then, when we return everything to the client, the application will filter out the comment that was requested?

a-w-m replied:
Got it.

rusye commented on src/routers/comment-router.js:
```diff
@@ -0,0 +1,95 @@
+  Comment.find({
+    $or: [{ userID: userID }, { text: text }, { date: date }]
+  })
+    .lean()
```
What is the `.lean()` supposed to do?

a-w-m replied:
According to the docs
> This makes queries faster and less memory intensive, but the result documents are plain old JavaScript objects (POJOs), not Mongoose documents. In this tutorial, you'll learn more about the tradeoffs of using lean().

The downside of enabling lean is that lean docs don't have:
Change tracking
Casting and validation
Getters and setters
Virtuals
save()

rusye replied:
Got it.

rusye commented on src/routers/comment-router.js:
```diff
@@ -0,0 +1,95 @@
+Sends JSON of comment back in response 
+*/
+
+router.post("/create", async function(req, res) {
```
You can leave the it as just `"/"`, no need for `/create`, only the `.post` can create

rusye commented on src/routers/comment-router.js:
```diff
@@ -0,0 +1,95 @@
+
+router.post("/create", async function(req, res) {
+  const { userID, text, date } = req.body;
+
```
You would want to check that a user exists before attempting to post a comment?

rusye commented on src/routers/comment-router.js:
```diff
@@ -0,0 +1,95 @@
+    },
+    function(err) {
+      if (err) {
+        res.status(500).send(err);
```
Are you sure the error will be a 500?

a-w-m replied:
I thought 500 error was for generic internal server error, i.e. the request was fine, but something went wrong on the server side.

rusye commented on src/routers/comment-router.js:
```diff
@@ -0,0 +1,95 @@
+    }
+  );
+
+   new_comment.populate("username", userID.username)
```
This `.populate()` would be ignored if you check that the user exists first and then store that info before you send it back

a-w-m replied:
Okay got it.

rusye commented on src/routers/comment-router.js:
```diff
@@ -0,0 +1,95 @@
+Takes request body and saves it as a comment in the database
+Populates username field once saved
+*/
+router.delete("/delete", function(req, res) {
```
You would want to pass through a variable in the url because you have it hard coded, so probably something like `.delete("/:postID", (req, res) => {...})`

rusye commented on src/routers/comment-router.js:
```diff
@@ -0,0 +1,95 @@
+router.delete("/delete", function(req, res) {
+  const { _id } = req.body;
+
+  Comment.findByIdAndDelete({ _id: _id }).exec(function(err) {
```
Is there a difference between `findByIdAndDelete` and `findByIdAndRemove`?

a-w-m replied:
They are very similar, interchangeable with most use cases according to the docs

>Query.prototype.findOneAndDelete()

This function differs slightly from Model.findOneAndRemove() in that findOneAndRemove() becomes a MongoDB findAndModify() command, as opposed to a findOneAndDelete() command. For most mongoose use cases, this distinction is purely pedantic. You should use findOneAndDelete() unless you have a good reason not to.

rusye commented on src/routers/comment-router.js:
```diff
@@ -0,0 +1,95 @@
+Receives an updated comment in request and updates comment in db
+Sends updated comment in response
+*/
+router.put("/edit", function(req, res) {
```
You could probably change it to a `/:postID` as well in here.

a-w-m committed:
Fixed comment routes

a-w-m committed:
Fixes comment routes

a-w-m committed:
Merged into dev

a-w-m committed:
Adds validation on post request comment route

a-w-m committed:
Updates Comment Routes, Schema and imports comment router to app.js

a-w-m committed:
Merge branch 'dev' into comment-router

Junjpan commented on src/routes/comments.js:
```diff
@@ -0,0 +1,122 @@
+      });
+
+      new_comment = await new_comment
+        .populate("username", userID.username)
```
I have a little confused in this part. Can you explain a little more details here?

a-w-m replied:
Yes, I can explain.  This routes post requests for comments. The second argument is an array that validates the contents of the request body, using an express-validator. The userID should be 24 characters in length and the text field should not have an empty value. The function returns a status error and array of errors generated by express-validator.

After confirming the request body is valid, I create a new mongoose ObjectID from userID. I confirm the user exists by passing the ObjectID into findOne(). If the user doesn't exist, I return 404 status error and an error message.

If the user exists, I create a new comment using the variables created by destructuring req.body. Once the new comment is created, I populate the username field using userID.

Let me know if it's still confusing, and/or how I can make it clearer or cleaner. Thanks!!

------

After reviewing this, I think I should use the mongoose ObjectID instead of userID when creating the comment:

 ` let new_comment = await Comment.create({
        userID: _id,
        text,
        date
      });
` 
The userID is a string, the mongoose ObjectID is an object. The string is converted into an object when used in find() queries, so I just wanted to be consistent and store the IDs as Objects instead of strings.

Junjpan replied:
In fact I originally referred to the codes below because this is not my normal way to do it. Anyway I'll pull your branch and test it.
 new_comment = await new_comment
        .populate("username", userID.username)
        .execPopulate();

Junjpan replied:
I just tested to delete the code below:

new_comment = await new_comment
.populate("username", userID.username)
.execPopulate();

the new_comment showed the same result with or without it.

rusye commented on src/app.js:
```diff
@@ -21,6 +23,10 @@ app.get("/", (req, res) => {
 app.use('/api/users', require('./routes/users'));
 
+// Create comment route
+app.use('/api/comments', require('./routes/comments'));
```
just curious, do we want to add the requirement up top like so? `const commentsRouter = require("...")`

rusye commented on src/routes/comments.js:
```diff
@@ -0,0 +1,122 @@
+  "/",
+  [
+    body("userID", "invalid userID").isLength({
+      min: 24
```
Can we really set a min length on this?  What if MongoDB changes things up?  Maybe also just have `.not().isEmpty()`?

rusye commented on src/routes/comments.js:
```diff
@@ -0,0 +1,122 @@
+    const _id = new mongoose.Types.ObjectId(userID);
+
+    try {
+      const user = await User.findOne({ _id });
```
Would `User.findById()` work better?

rusye commented on src/routes/comments.js:
```diff
@@ -0,0 +1,122 @@
+    try {
+      const user = await User.findOne({ _id });
+      if (!user) {
+        return res.status(404).json({ error: "No user Found" });
```
Would a 400 be more accurate? https://httpstatuses.com/400

rusye commented on src/routes/comments.js:
```diff
@@ -0,0 +1,122 @@
+        date
+      });
+
+      new_comment = await new_comment
```
Do we need this?  All the user information is already stored in the `user` on line 59, therefore could you send back something like this ```return res.json({
user: {...user},
...new_comment //from line 64
})```?

Junjpan commented on src/routes/comments.js:
```diff
@@ -0,0 +1,122 @@
+router.get("/", function(req, res) {
+  Comment.find()
+    .lean()
+    .exec(function(err, comments) {
```
I think you should do the populate here and get the username. Of course, you can do that later as well.

a-w-m replied:
Great, I went ahead and moved it here.

a-w-m committed:
Fixes Comment Router

-Moves populate to get route
- tweak get response
- update error handling
- remove callbacks
-insert then/catch
-fix http response error status
-change Schema field from userID to user
-move router require up in app.js

rusye commented on src/models/comment.js:
```diff
@@ -4,9 +4,9 @@ const mongoose = require("mongoose");
 
 const commentSchema = new mongoose.Schema({
-  userID: { type: Schema.Types.ObjectId, ref: "User" },
+  user: { type: mongoose.Schema.Types.ObjectId, ref: User },
```
I believe you still need quotes around User, `ref: "User"`

a-w-m replied:
For some reason the quotes don't work on my end. It works without the quotes though. Not sure why.

rusye commented on src/routes/comments.js:
```diff
@@ -0,0 +1,118 @@
+        comments.map(comment => {
+          return {
+            _id: comment._id,
+            user: comment.user._id,
```
The front end doesn't need to see the `user._id`, so initially we have the user be the user id, after we get the data from `populate()` we replace the id with the user name

a-w-m replied:
Okay got it, I fixed it so that user returns the username in the response, and deleted user._id

a-w-m committed:
Remove userID from get response
